firmware/application/tone_key.cpp

```diff
@@ -114,10 +114,58 @@ std::string tone_key_string(tone_index index) {
     return tone_keys[index].first;
 }
 
-std::string tone_key_string_by_value(uint32_t value) {
-    return tone_key_string(tone_key_index_by_value(value));
+// Return string showing frequency only from specific table index
+std::string tone_key_value_string(tone_index index) {
+    if (index < 0 || (unsigned)index >= tone_keys.size())
+        return std::string("");
+    auto f = tone_keys[index].second;
+    return to_string_dec_uint((uint32_t)f) + "." + to_string_dec_uint((uint32_t)(f * 10) % 10);
+}
+
+// Return variable-length string showing CTCSS tone from tone frequency
+// Value is in 0.01 Hz units
+std::string tone_key_string_by_value(uint32_t value, size_t max_length) {
+    static uint8_t tone_display_toggle{0};
+    static uint32_t last_value;
+    tone_index idx;
+    std::string freq_str;
+
+    // If >10Hz difference between consecutive samples, it's probably noise, so ignore
+    if (abs(value - last_value) > 10 * 100) {
+        last_value = value;
+        tone_display_toggle = 0;
+        return "        ";
+    }
+    last_value = value;
+
+    // Only display 1/10 Hz accuracy if <1000 Hz; max 5 characters
+    if (value >= 1000 * 100)
+        freq_str = to_string_dec_int(value / 100);
+    else
+        freq_str = to_string_dec_int(value / 100) + "." + to_string_dec_uint((value / 10) % 10);
+
+    if (max_length >= 3 + 5 + 1 + 5) {
+        return "T: " + freq_str + " " + tone_key_string(tone_key_index_by_value(value));
+    } else {
+        // Not enough space; toggle between display of tone received and tone code #
+        if (tone_display_toggle++ < TONE_DISPLAY_TOGGLE_COUNTER) {
+            return "T: " + freq_str;
+        } else {
+            if (tone_display_toggle >= TONE_DISPLAY_TOGGLE_COUNTER * 2)
```

Comment:
TONE_DISPLAY_TOGGLE_COUNTER * 2?
I'm trying to understand these constants. Toggle sounds like a bool, but you're using values like 2 and 4?

Reply:
It's just one up-counter from 0 to TONE_DISPLAY_TOGGLE_COUNTER*2 and the display string flips halfway in between.  I thought the name was OK but I'll be happy to change it to whatever.

Reply:
I was just wondering why you count from 0 to 3 instead of 0 to 1.